### PR TITLE
Implementation of the Shadow Thing mechanism on the device side

### PR DESCRIPTION
**Pull request by ilcato**
It allows the synchronization of property values between cloud and devices, at startup or after a disconnection period, based on synchronization policies configurable at the property level.
It adds a new parameter to the `addProperty` method that specify the synchronization mode to adopt for the property. This parameter is a callback function that will specify the synchronization logic to use. Standard synchronization logic is available by using the following callbacks:
* AUTO_SYNC: that compares the **cloud** timestamp of the last change with the corresponding **device** timestamp. The property is assigned the value with the higher timestamp.
* FORCE_CLOUD_SYNC: the property is assigned the value coming from **cloud** regardless of timestamps and **device** value. 
* FORCE_DEVICE_SYNC: the device property value is kept. The cloud property value will be updated at the next update cycle.

A custom synchronization logic may be implemented by setting a custom callback function with the following signature: `void (*)(ArduinoCloudProperty<T> property)`; use one of the specific types supported. The property object exposes several methods which will enable the custom logic to select the appropriate value.

Example: assuming a Switch boolean READ_WRITE property is defined the following statement in `thingProperties.h` will specify an AUTO_SYNC mode:

`ArduinoCloud.addProperty(Switch, READWRITE, ON_CHANGE, onSwitchChange, AUTO_SYNC);`

### CODE REVIEW

File: examples/shadow_thing_sync_callbacks.ino [hunk 1]
```diff
@@ -0,0 +1,103 @@
+#include "arduino_secrets.h"
+
+<<<<<<< HEAD
```
Comment (lxrobotics) [inline examples/shadow_thing_sync_callbacks.ino]: It appears that there are merge conflict artifacts left within the whole file - please resolve.
```suggestion
```
Overall I appreciate the checking in of an example. Regardless - this example requires the **ArduinoIoTCloud** and therefore I suggest moving the example over to the ArduinoIoTCloud library.

File: examples/thingProperties.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#include <ArduinoIoTCloud.h>
+#include <WiFiConnectionManager.h>
+
+char ssid[] = SECRET_SSID;    // Network SSID (name)
+char pass[] = SECRET_PASS;    // Network password (use for WPA, or use as key for WEP)
+
+#define THING_ID "PUT_YOUT_THING_ID_HERE"
+
+void onSwitchChange();
+<<<<<<< HEAD
```
Comment (lxrobotics) [inline examples/thingProperties.h]: It seems that there are more merge conflicts in here - please resolve.
```suggestion
```

File: src/ArduinoCloudThing.cpp [hunk 3]
```diff
@@ -460,8 +497,8 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
 bool ArduinoCloudThing::ifNumericConvertToDouble(CborValue * value_iter, double * numeric_val) {
 
   if(cbor_value_is_integer(value_iter)) {
-    int val = 0;
-    if(cbor_value_get_int(value_iter, &val) == CborNoError) {
+    uint64_t val = 0;
```
Comment (lxrobotics) [inline src/ArduinoCloudThing.cpp]: I'm not sure that this is correct. If we retrieve an **unsigned int** instead of a **signed int** we are loosing the sign extension. As a consequence **no negative property values would be possible**.

I'd be great if you could write up some decode unit tests where CBOR streams containing negative values are passed. As of now none do exist and therefore this error would have gone unnoticed by the test suite.

File: src/ArduinoCloudThing.h [hunk 4]
```diff
@@ -142,6 +159,7 @@ class ArduinoCloudThing {
   MapParserState handle_Time         (CborValue * value_iter, CborMapData * map_data);
   MapParserState handle_LeaveMap     (CborValue * map_iter, CborValue * value_iter, CborMapData const * const map_data);
 
+  static void   resetMapDataNotBase         (CborMapData * map_data);
```
Comment (lxrobotics) [inline src/ArduinoCloudThing.h]: I'd suggest function `resetMapDataNotBase` should better be a member function of the class `CborMapData` - as the `reset` function was.

File: src/ArduinoCloudThing.h [hunk 5]
```diff
@@ -40,6 +40,31 @@ static long const MINUTES   = 60;
 static long const HOURS     = 3600;
 static long const DAYS      = 86400;
 
+/******************************************************************************
+ * SYNCHRONIZATION CALLBACKS
+ ******************************************************************************/
+
+#define AUTO_SYNC onAutoSync
```
Comment (lxrobotics) [inline src/ArduinoCloudThing.h]: Is there any reason for using those defines? Stroustrop's opinion on macros (somewhat simplified by me): Don't use them unless you have to ( http://www.stroustrup.com/bs_faq2.html#macro ). If you really want the function call to look like `AUTO_SYNC` then I'd suggest you just define it that way `void AUTO_SYNC(ArduinoCloudProperty<T> property)`.

File: src/ArduinoCloudThing.cpp [hunk 6]
```diff
@@ -491,6 +528,14 @@ bool ArduinoCloudThing::ifNumericConvertToDouble(CborValue * value_iter, double
   return false;
 }
 
+void ArduinoCloudThing::resetMapDataNotBase(CborMapData * map_data) {
```
Comment (lxrobotics) [inline src/ArduinoCloudThing.cpp]: I'd see that function as a member of class **CborMapData** as I've noted in the header file too.

**Comment by lxrobotics** [comment]
Also please perform a rebase incorporating the current state of master in order to pull in the (minimal) changes done to testing code.

File: examples/shadow_thing_sync_callbacks.ino (hunk 1 above)
Comment (ilcato) [reply examples/shadow_thing_sync_callbacks.ino]: It was an old example that I put here erroneously.

File: examples/thingProperties.h (hunk 2 above)
Comment (ilcato) [reply examples/thingProperties.h]: It was an old example that I put here erroneously.

File: src/ArduinoCloudThing.h (hunk 5 above)
Comment (ilcato) [reply src/ArduinoCloudThing.h]: I'd prefer to keep the function name as is. The defines are there in order to align with the way in which in the auto generated thingProperties.h the UI configurable properties characteristics are defined. See macro [here](https://github.com/ilcato/ArduinoIoTCloud/blob/d532185eaa4c523458361fb0592467d1c4ff29d7/src/ArduinoIoTCloud.h#L91).
A sample property declaration is:
`  ArduinoCloud.addProperty(Switch, READWRITE, ON_CHANGE, onSwitchChange, AUTO_SYNC);
`

File: src/ArduinoCloudThing.cpp (hunk 3 above)
Comment (ilcato) [reply src/ArduinoCloudThing.cpp]: Correct, I fixed it by declaring a correct type and corresponding cobra library function call. I also added a specific test for negative case. The reason I change the type from int to int64 is that there are cases in which a 32 bit integers doesn't fit the values to manage.

File: src/ArduinoCloudThing.cpp (hunk 6 above)
Comment (ilcato) [reply src/ArduinoCloudThing.cpp]: Done

File: src/ArduinoCloudThing.h (hunk 4 above)
Comment (ilcato) [reply src/ArduinoCloudThing.h]: Done

File: README.md [hunk 7]
```diff
@@ -32,7 +32,15 @@ Additional configuration can be added to the property via composition
 
 `.publishOnChange()`
 
-`.onUpdate(functionPointer)` configures the property to call `functionPointer` when the value is changed in the cloud.
+`.onUpdate(functionPointer)` configures the property to call `functionPointer` when the value is changed by the cloud.
+
+`.onSync(syncFunctionPointer)` configures the property to call `syncFunctionPointer` when as consequence of a connection/reconnection (so after a period passed offline) the value of the property must be synchronized with the one stored in the cloud. The implemetation `syncFunctionPointer()` should contain some synchornization logic; standard implementations for this functions are:
```
Comment (per1234) [inline README.md]:
```suggestion
`.onSync(syncFunctionPointer)` configures the property to call `syncFunctionPointer` when as consequence of a connection/reconnection (so after a period passed offline) the value of the property must be synchronized with the one stored in the cloud. The implementation `syncFunctionPointer()` should contain some synchronization logic; standard implementations for this functions are:
```

File: README.md [hunk 8]
```diff
@@ -43,17 +51,17 @@ void onUpdateCallback() {
...
+thing.addProperty(int_property, "test_int_property", Permission::ReadWrite).publishOnChange(min_delta).onUpdate(onUpdateCallback).onSync(onSynchronizationCallback);
 ...
 ```
 
 * **encode** Checks all properties if their value has changed and if so populates the buffer with the desired CBOR update message.
 
 `int encode(uint8_t * data, size_t const size)`
 
-* **decode** decodes a CBOR buffer received from the cloud and updates writeable properties accordingly. Also the update callbacks are called, if the value of a property has changed.
+* **decode**  decodes a CBOR buffer received from the cloud and if the syncMessage parameter is set to false (default), updates writeable properties accordingly and call the update callback if the value of a property has changed. If the syncMessage parameter is set to true, the value of the property received by the cloud and the relative last change timestamp are passed to a synchronization callback. The synchronization callback could apply the logic to assign a value to the property choosing between the local value and the cloud one. The synchronization logic also decide if the call to the update callback is necessary.
```
Comment (per1234) [inline README.md]:
```suggestion
* **decode**  decodes a CBOR buffer received from the cloud and if the syncMessage parameter is set to false (default), updates writeable properties accordingly and calls the update callback if the value of a property has changed. If the syncMessage parameter is set to true, the value of the property received by the cloud and the relative last change timestamp are passed to a synchronization callback. The synchronization callback could apply the logic to assign a value to the property choosing between the local value and the cloud one. The synchronization logic also decides if the call to the update callback is necessary.
```

File: src/ArduinoCloudProperty.hpp [hunk 9]
```diff
@@ -91,26 +92,48 @@ class ArduinoCloudProperty {
...
+                     _local_shadow_property;
   String             _name;
   Permission         _permission;
   UpdateCallbackFunc _update_callback_func;
+  void               (*_sync_callback_func)(ArduinoCloudProperty<T> property);
 
   UpdatePolicy       _update_policy;
   bool               _has_been_updated_once,
-                     _has_been_modified_in_callback;
+                     _has_been_modified_in_callback;
```
Comment (per1234) [inline src/ArduinoCloudProperty.hpp]:
```suggestion
                     _has_been_modified_in_callback;
```

File: src/ArduinoCloudThing.cpp [hunk 10]
```diff
@@ -404,6 +409,14 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_Time(CborValue * val
 ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue * map_iter, CborValue * value_iter, CborMapData const * const map_data) {
   MapParserState next_state = MapParserState::Error;
 
+  //computer the cloud event change time
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
  //compute the cloud event change time
```

File: src/ArduinoCloudThing.cpp [hunk 11]
```diff
@@ -414,31 +427,55 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
       ArduinoCloudProperty<float> * float_property = _property_cont.getPropertyFloat(map_data->name.get());
 
       if(int_property && int_property->isWriteableByCloud()) {
-        int_property->writeByCloud(static_cast<int>(map_data->val.get())); /* Val is internally stored as float */
-        int_property->execCallbackOnChange();
+        if(_syncMessage){
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
        if(_syncMessage) {
```

File: src/ArduinoCloudThing.cpp [hunk 12]
```diff
@@ -414,31 +427,55 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+          int_property->setLastCloudChangeTimestamp(cloudChangeEventTime);
+          int_property->setCloudShadowValue(map_data->val.get());
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
          int_property->setCloudShadowValue(map_data->val.get());
```

File: src/ArduinoCloudThing.cpp [hunk 13]
```diff
@@ -414,31 +427,55 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+        } else {
+          int_property->writeByCloud(map_data->val.get());
+          int_property->execCallbackOnChange();
+        }
       }
 
       if(float_property && float_property->isWriteableByCloud()) {
-        float_property->writeByCloud(map_data->val.get());
-        float_property->execCallbackOnChange();
+        if(_syncMessage){
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
        if(_syncMessage) {
```

File: src/ArduinoCloudThing.cpp [hunk 14]
```diff
@@ -414,31 +427,55 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+          float_property->setLastCloudChangeTimestamp(cloudChangeEventTime);
+          float_property->setCloudShadowValue(map_data->val.get());
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
          float_property->setCloudShadowValue(map_data->val.get());
```

File: src/ArduinoCloudThing.cpp [hunk 15]
```diff
@@ -414,31 +427,55 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
       }
     }
 
     /* Value (String) */
     if(map_data->str_val.isSet()) {
       ArduinoCloudProperty<String>* string_property = _property_cont.getPropertyString(map_data->name.get());
       if(string_property && string_property->isWriteableByCloud()) {
-        string_property->writeByCloud(map_data->str_val.get());
-        string_property->execCallbackOnChange();
+        if(_syncMessage){
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
        if(_syncMessage) {
```

File: src/ArduinoCloudThing.cpp [hunk 16]
```diff
@@ -414,31 +427,55 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+          string_property->setLastCloudChangeTimestamp(cloudChangeEventTime);
+          string_property->setCloudShadowValue(map_data->str_val.get());
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
          string_property->setCloudShadowValue(map_data->str_val.get());
```

File: src/ArduinoCloudThing.cpp [hunk 17]
```diff
@@ -414,31 +427,55 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
       }
     }
 
     /* Value (bool) */
     if(map_data->bool_val.isSet()) {
       ArduinoCloudProperty<bool>* bool_property = _property_cont.getPropertyBool(map_data->name.get());
       if(bool_property && bool_property->isWriteableByCloud()) {
-        bool_property->writeByCloud(map_data->bool_val.get());
-        bool_property->execCallbackOnChange();
+        if(_syncMessage){
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
        if(_syncMessage) {
```

File: src/ArduinoCloudThing.cpp [hunk 18]
```diff
@@ -414,31 +427,55 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+          bool_property->setLastCloudChangeTimestamp(cloudChangeEventTime);
+          bool_property->setCloudShadowValue(map_data->bool_val.get());
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
          bool_property->setCloudShadowValue(map_data->bool_val.get());
```

File: src/ArduinoCloudProperty.ipp [hunk 19]
```diff
@@ -89,19 +100,33 @@ bool ArduinoCloudProperty<T>::shouldBeUpdated() {
...
 template <typename T>
 void ArduinoCloudProperty<T>::execCallbackOnChange() {
-  if(isValueDifferent(_property, _shadow_property)) {
+  if(isValueDifferent(_property, _cloud_shadow_property)) {
     if(_update_callback_func != NULL) {
       _update_callback_func();
     }
-
-    if(!isValueDifferent(_property, _shadow_property)) {
+
```
Comment (per1234) [inline src/ArduinoCloudProperty.ipp]:
```suggestion

```

File: test/src/test_callback.cpp [hunk 20]
```diff
@@ -113,3 +113,174 @@ SCENARIO("A (boolean) property is manipulated in the callback to its origin stat
...
+  MOST_RECENT_WINS(property);
+  sync_callback_called = true;
+}
+
+void change_callback()
+{
+  change_callback_called = true;
+}
+
+SCENARIO("After a connection/reconnection an incoming cbor payload is processed and the synchronization callback is executed. The sync callback apply the AUTO_SYNC policy (the most recent value between the local one and the cloud one is finally assigned to the property). The onUpdate function is called if the cloud value is the most recent one. In this scenario the most updated value is the cloud one.")
```
Comment (per1234) [inline test/src/test_callback.cpp]:
```suggestion
SCENARIO("After a connection/reconnection an incoming cbor payload is processed and the synchronization callback is executed. The sync callback applies the AUTO_SYNC policy (the most recent value between the local one and the cloud one is finally assigned to the property). The onUpdate function is called if the cloud value is the most recent one. In this scenario the most updated value is the cloud one.")
```

File: test/src/test_callback.cpp [hunk 21]
```diff
@@ -113,3 +113,174 @@ SCENARIO("A (boolean) property is manipulated in the callback to its origin stat
...
+
+/**************************************************************************************/
+
+void force_device_sync_callback(ArduinoCloudProperty<bool> property)
+{
+  DEVICE_WINS(property);
+  sync_callback_called = true;
+}
+
+SCENARIO("After a connection/reconnection an incoming cbor payload is processed and the synchronization callback is executed. The sync callback apply the FORCE_DEVICE_SYNC policy (the property keep the local value and, if the cloud value is different from the local one, the value is propagated to the cloud). The onUpdate function is not executed")
```
Comment (per1234) [inline test/src/test_callback.cpp]:
```suggestion
SCENARIO("After a connection/reconnection an incoming cbor payload is processed and the synchronization callback is executed. The sync callback applies the FORCE_DEVICE_SYNC policy (the property keeps the local value and, if the cloud value is different from the local one, the value is propagated to the cloud). The onUpdate function is not executed")
```

File: test/src/test_callback.cpp [hunk 22]
```diff
@@ -113,3 +113,174 @@ SCENARIO("A (boolean) property is manipulated in the callback to its origin stat
...
+
+/**************************************************************************************/
+
+void force_cloud_sync_callback(ArduinoCloudProperty<bool> property)
+{
+  CLOUD_WINS(property);
+  sync_callback_called = true;
+}
+
+SCENARIO("After a connection/reconnection an incoming cbor payload is processed and the synchronization callback is executed. The sync callback apply the FORCE_CLOUD_SYNC policy (the property always assume the value incoming from the broker message). The onUpdate function is executed only if the local value of the property was different from the one taken from the incoming message")
```
Comment (per1234) [inline test/src/test_callback.cpp]:
```suggestion
SCENARIO("After a connection/reconnection an incoming cbor payload is processed and the synchronization callback is executed. The sync callback applies the FORCE_CLOUD_SYNC policy (the property always assumes the value incoming from the broker message). The onUpdate function is executed only if the local value of the property was different from the one taken from the incoming message")
```

File: test/src/test_callback.cpp [hunk 23]
```diff
@@ -113,3 +113,174 @@ SCENARIO("A (boolean) property is manipulated in the callback to its origin stat
...
+    REQUIRE(sync_callback_called == true);
+    REQUIRE(change_callback_called == true);
+
+    REQUIRE(test == true);
+  }
+}
+
+/**************************************************************************************/
+
+SCENARIO("After a connection/reconnection an incoming cbor payload is processed. Any synchoniation function is passed to the property so the value in the incoming message is discarded")
```
Comment (per1234) [inline test/src/test_callback.cpp]:
```suggestion
SCENARIO("After a connection/reconnection an incoming cbor payload is processed. Any synchronization function is passed to the property so the value in the incoming message is discarded")
```

**Comment by lxrobotics** [comment]
@ilcato Thank you very much for incorporating @per1234 suggestions. May I ask you to perform a rebase and squash all those formatting commits into a single one in order to unpollute the commit history?

**Comment by per1234** [comment]
@ilcato thanks so much for considering my suggestions. lxrobotics requested that I take another look at the PR to see if this now meets my approval but I notice that some of my suggestions are still unresolved. You're not required to accept them of course, but it would be nice to add a comment explaining the reasoning for any you want to reject. I don't think any of the the outstanding ones are controversial.
